turbine/src/addr_cache.rs

```diff
@@ -0,0 +1,412 @@
+    index_data: usize,
+    // Maximum code and data indices observed in retransmit-stage.
+    max_index_code: u32,
+    max_index_data: u32,
+    // If the last data shred in the slot is already observed, the cache is no
+    // longer extended beyond max_index_{code,data}.
+    last_shred_in_slot: bool,
+}
+
+impl AddrCache {
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            capacity,
+            num_shreds: 0,
+            window: VecDeque::new(),
+            counts: HashMap::new(),
+            cache: {
+                // 2x capacity in order to implement lazy eviction.
+                let capacity = capacity.saturating_mul(2).saturating_add(1);
+                HashMap::with_capacity(capacity)
+            },
+        }
+    }
+
+    // Returns (root-distance, socket-addresses) cached for the given shred-id.
+    #[inline]
+    pub(crate) fn get(&self, shred: &ShredId) -> Option<(/*root_distance:*/ u8, &[SocketAddr])> {
+        self.cache
+            .get(&shred.slot())?
+            .get(shred.shred_type(), shred.index())
+    }
+
+    // Stores (root-distance, socket-addresses) precomputed speculatively for
+    // the given shred-id.
+    pub(crate) fn put(
+        &mut self,
+        shred: &ShredId,
+        entry: (/*root_distance:*/ u8, Box<[SocketAddr]>),
+    ) {
+        self.get_cache_entry_mut(shred.slot())
+            .put(shred.shred_type(), shred.index(), entry);
+        self.maybe_trim_cache();
+    }
+
+    // Records data observed from incoming shreds at retransmit stage.
+    pub(crate) fn record(&mut self, slot: Slot, stats: &mut RetransmitSlotStats) {
+        // All addresses should be for the same slot.
+        debug_assert!(stats.addrs.iter().all(|(shred, _, _)| shred.slot() == slot));
+        // Update rolling window count of shreds per slot.
+        let num_shreds: usize = stats.num_shreds_received.iter().sum();
+        if num_shreds > 0 {
+            self.num_shreds += num_shreds;
+            self.window.push_back((slot, num_shreds));
+            *self.counts.entry(slot).or_default() += num_shreds;
+            self.maybe_trim_slot_counts();
+        }
+        debug_assert!(self.verify());
+        // If there are no addresses to cache and the cache entry is not
+        // allocated for the slot yet, then ignore.
+        if stats.addrs.is_empty() && !self.cache.contains_key(&slot) {
+            return;
+        }
+        // Update the cached entry for the slot.
+        let entry = self.get_cache_entry_mut(slot);
+        entry.max_index_code = entry.max_index_code.max(stats.max_index_code);
+        entry.max_index_data = entry.max_index_data.max(stats.max_index_data);
+        entry.last_shred_in_slot |= stats.last_shred_in_slot;
+        for (shred, root_distance, addrs) in std::mem::take(&mut stats.addrs) {
+            debug_assert_eq!(shred.slot(), slot);
+            entry.put(shred.shred_type(), shred.index(), (root_distance, addrs));
```

Comment:
At this point why do we need to insert the shred into the cache? `record` is called after we've retransmitted the shred, so barring a duplicate shred we should not expect having to retransmit again right?

Reply:
yeah, it is just if there is a duplicate shred.